criu/kerndat.c

```diff
@@ -1055,64 +1080,124 @@ int kerndat_init(void)
+		ret = 1;
+	}
+	if (!ret && init_zero_page_pfn()) {
+		pr_err("Failed to initialize zero page page frame number when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && get_last_cap()) {
+		pr_err("Failed to read last cap when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_fdinfo_has_lock()) {
+		pr_err("Failed to get lock in fdinfo when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && get_task_size()) {
+		pr_err("Failed to get task size when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && get_ipv6()) {
+		pr_err("Failed to access internet protocol when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_loginuid()) {
+		pr_err("Failed to update loginuid when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_iptables_has_xtlocks()) {
+		pr_err("Failed to get xlocks in iptables when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_tcp_repair()) {
+		pr_err("Failed to set socket to TCP_REPAIR when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_compat_restore()) {
+		pr_err("Failed to update compatible restore when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_tun_netns()) {
+		pr_err("Failed to get net namespace in tuns when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_socket_unix_file()) {
+		pr_err("Failed to create a socket when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_nsid()) {
+		pr_err("Failed to get namespace ID when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_link_nsid()) {
+		pr_err("Failed to get link NSID when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_has_memfd_create()) {
+		pr_err("Failed to create memfd when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_detect_stack_guard_gap()) {
+		pr_err("Failed to set up stack guard when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_uffd()) {
+		pr_err("Failed to open userfaultfd when initializing kerndat.\n");
+		ret = 1;
+	}
+	if (!ret && kerndat_has_thp_disable()) {
+		pr_err("Failed to prctl PR_SET_THP_DISABLE when initializing kerndat.\n");
+		ret = 1;
+	}
 	/* Needs kdat.compat_cr filled before */
-	if (!ret)
-		ret = kerndat_vdso_fill_symtable();
+	if (!ret && kerndat_vdso_fill_symtable()) {
+		pr_err("Failed to fill self vdso symtable when initializing kerndat\n");
+		ret = 1;
+	}
 	/* Depends on kerndat_vdso_fill_symtable() */
-	if (!ret)
-		ret = kerndat_vdso_preserves_hint();
-	if (!ret)
-		ret = kerndat_socket_netns();
-	if (!ret)
-		ret = kerndat_x86_has_ptrace_fpu_xsave_bug();
-	if (!ret)
-		ret = kerndat_has_inotify_setnextwd();
-	if (!ret)
-		ret = has_kcmp_epoll_tfd();
-	if (!ret)
-		ret = kerndat_has_fsopen();
-	if (!ret)
-		ret = kerndat_has_clone3_set_tid();
-	if (!ret)
-		ret = has_time_namespace();
+	if (!ret && kerndat_vdso_preserves_hint()) {
+		pr_err("Failed to parse vDSO map from child helper when initializing kerndat.\n");
```

Comment:
I am especially not sure about how to best log this function if an error occurs. I do not entirely comprehend how the forking within this function "preserves hint."